TypeCobol.Analysis/Graph/CfgAbstractInterpretation.Environment.cs

```diff
@@ -0,0 +1,341 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using TypeCobol.Analysis.Util;
+
+namespace TypeCobol.Analysis.Graph
+{
+    public partial class CfgAbstractInterpretation<N,D>
+    {
+        /// <summary>
+        /// Abstract interpretation execution environment.
+        /// The default execution performed is based on a dfs algorithm, with exception that
+        /// branching instructions are handled specifically.
+        /// Thus not all execution paths are applied, but all nodes are visited using dfs like algorithm if they are accessible.
+        /// A node cannot be visited twice.
+        /// </summary>
+        public class Environment
+        {
+            /// <summary>
+            /// The Abstract Interpretation Stack
+            /// </summary>
+            private Stack<BasicBlock<N, D>> _interpretationStack;        
+
+            /// <summary>
+            /// The current execution stack
+            /// </summary>
+            public IReadOnlyCollection<BasicBlock<N, D>> ExecutionStack => _interpretationStack;
+
+            /// <summary>
+            /// The DFS Run stack
+            /// </summary>
+            private Stack<BasicBlock<N, D>> _dfsStack;            
+            /// <summary>
+            /// Bit Vector to check if metric values have been taken in account for a block index.
+            /// </summary>
+            private System.Collections.BitArray _metricUpdated;
+            /// <summary>
+            /// Current Cyclic execution Thresold by block index.
+            /// </summary>
+            private Dictionary<int,int> _cyclicThreshold;
+            /// <summary>
+            /// The Cyclic relation from one block index to another block index
+            /// </summary>
+            private Dictionary<int, HashSet<int> >_cyclicRelation;
```

Comment:
The _cyclicRelation is only for remembering which relation (which transition) leads to a cycle, it is not the entire transtive closure relation of the graph. So it now called _cyclicTransition.
We just want to limit how many time a cyclic transition will be executed to avoid infinite transitions. otherwise there is no limitation on the number of time a non cyclic transition can be executed.

Reply:
I understood that, I know we have to limit the number of times a block is visited. My idea is to let the client decide instead  of fixing a maximum for all blocks.

Reply:
How can client decides on block he doesn't know? the cyclicExecutionThreshold is already a parameter of the method, do you mean that we should add an option for that? even more we don't know in adance what transition will be cyclic.
So there can only be an unique threshold for each block.